File: subsets/subsets.cpp
class Solution {
public:
    void solve(vector<int> input, vector<int> & output, vector<vector<int>> &op)
    {
        if(input.size()==0)
        {
            vector<int> sum;
            for(auto it: output)
                sum.push_back(it);
            op.push_back(sum);
            return;
        }
        vector<int> out1=output;
        vector<int> out2=output;
        
        out2.push_back(input[0]);
        input.erase(input.begin()+0);
        solve(input,out1, op);
        solve(input,out2, op); 
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> output;
        vector<vector<int>> out;
        solve(nums, output, out);
        return out;
    }
};

File: subsets/subsets.java
class Solution {
    List<List<Integer>> fin = new ArrayList<List<Integer>>();
    
    public List<List<Integer>> subsets(int[] nums) {
        int i= 0;
        ArrayList<Integer> curr = new ArrayList<Integer>();
        ArrayList<Integer> array_list = new ArrayList<Integer>();
        for (int j = 0; j < nums.length; j++)
        {
            array_list.add(new Integer(nums[j]));
        }
        subsetsolver(array_list, new ArrayList<>(),0);
        return fin;
    }
    public void subsetsolver( ArrayList<Integer> nums , ArrayList<Integer> curr, int i) {
        if(i==nums.size())
        {
            ArrayList<Integer> out = new ArrayList<Integer>();
            for(int j=0;j<curr.size();j++)
                out.add(curr.get(j));
                //System.out.print(curr.get(j));
            //System.out.println();
            fin.add(out);
            return;
        }
        
        subsetsolver(nums, curr,i+1);
        curr.add(nums.get(i));
        subsetsolver(nums, curr,i+1);
        curr.remove(curr.size()-1);
        
    }
}